scripts/link_linter.js

```diff
@@ -0,0 +1,191 @@
+    const file = files[0];
+    const filePath = path.resolve(__dirname, file);
+
+    // Get the "stat" for the file to check if it is a directory.
+    const fstat = fs.statSync(filePath);
+
+    // Split the file string to capture the file suffix.
+    const fileParts = file.split(".");
+    const fileSuffix = fileParts[fileParts.length - 1];
+
+    // If the file is a directory, check the contents of the directory.
+    if (fstat.isDirectory()) {
+        const dirFiles = fs.readdirSync(filePath).map(function(file) {
+            return path.join(filePath, file);
+        });
+        files[0] = dirFiles;
+
+        // Add the contents of the directory to the original file list.
+        const combinedFiles = [].concat.apply([], files);
+        return fetchFilePaths(combinedFiles, result);
+    }
+
+    // If the file has a "html" suffix add it the result list.
+    if (fileSuffix.toLowerCase() === "html") {
+        // Check for autogenerated files.
+        const htmlString = fs.readFileSync(filePath, "utf8");
+        const autoGeneratedString = "<!-- WARNING: this was generated by a tool. Do not edit it by hand. -->";
+
+        if (htmlString.indexOf(autoGeneratedString) === -1) {
+            result.push(filePath);
+        }
+    }
+    const remainingFiles = files.slice(1, files.length);
+
+    // If there are files remaining check them.
+    if (remainingFiles.length > 0) {
+        return fetchFilePaths(remainingFiles, result);
+    }
+
+    // Return the resulting list of html files.
+    return result;
+}
+
+/**
+ * Grab the initial list of files in the defined directory and pass them
+ * to recursive function to build the file list.
+ *
+ * @param {string} directoryPath The path relative to the script where the files that need to be checked live.
+*/
+function buildFileList(directoryPath) {
+    const files = fs.readdirSync(path.resolve(__dirname, directoryPath)).map(function(file) {
+        return path.join(directoryPath, file);
+    });
+
+    return fetchFilePaths(files, []);
+}
+
+/**
+ * Find the line number of the first occurance of a string in a large string (i.e. html).
+ * Returns 0 if the string is not found.
+ *
+ * @param {string} blockText The block of text you want to find a strings line number for.
+ * @param {string} string The string you want to find the line number for.
+ */
+function getLinePosition(blockText, string) {
+    // Find the starting index of the string in the block.
+    const stringIndex = blockText.indexOf(string);
+
+    // Return zero if the string is not found.
+    if (stringIndex === -1) {
+        return 0;
+    }
+
+    // Grab everything leading up the start of the specified string.
+    const shortenedBlock = blockText.substr(0, stringIndex);
+
+    // Split the shortened block by '\n' and return the length of the array.
+    return shortenedBlock.split("\n").length;
+}
+
+const DATA_TRACKING_ID_REGEX = /data-tracking-id=("|')([a-zA-Z0-9]+)([-{ ()$.a-zA-Z0-9}]+)("|')/g;
```

Comment:
Is there any reason why we don't want to only allow `[a-zA-Z0-9_-.]+`? Allowing `{ ()` seems like it's just asking for trouble, since we'd be at the mercy of how whatever 3rd party system interprets/honors the data. (e.g. GA might URL encode the value, replacing " " with "%20" automatically.

It's a pain to have a stricter format for these tracking IDs, but since we control the IDs put in our markup, it's probably better to start strict now since it would be harder to change later.

Reply:
In the current state of the linter, the files are linted pre-compile. So before `hugo` does its magic. This means there will be ids like `shortcodes-glossary-{{ .term | urlize }}`, `shortcodes-changlelog-row-{{ $version | urlize }}`, `home-article-{{ (index .Params.articles 1).title | urlize }}`, etc...

By using `hugo` variables we could get rid of most of the symbols, and in that case we just have to include `{}` and `$` as symbols in the regex. 

So I think the best solution is just to add rules to validate the IDs are in the correct format.